lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -572,6 +572,8 @@ Type TypeChecker::typeCheckParameterDefault(Expr *&defaultValue,
 
   auto paramInterfaceTy = paramType->mapTypeOutOfContext();
 
+  auto initialTypecheckQueue = DiagnosticQueue(ctx.Diags,
```

Comment:
Both the original and new approach are incorrect because we might evaluate a request evaluator request which will emit diagnostics the first time we type check, but then the result is already cached the second time so we will lose this diagnostic if that first type check fails. Let's find a way to do this without diagnostic transactions.

CC: @xedin

Reply:
@slavapestov thanks for the feedback. i'm a little unclear on the scenario you're highlighting... are you saying we might process a request evaluator request after this point (within the call to `typeCheckExpression`), or that this code may be run as part of such a request (or both)? in any case, i'm not following why the diagnostics would be lost – are they persisted differently for request evaluator requests than they otherwise would be?

and to further clarify – when you say:

> Let's find a way to do this without diagnostic transactions.

by 'this', do you mean 'type check the parameter default' and not 'fix/bandaid the downstream SILGen crash'? i can think of some more minimal somewhat hacky solutions to the latter, but don't have a great sense of where to start with the former (any ideas would be appreciated).

Reply:
I agree with @slavapestov, use of transactions is very unfortunate here, I was hoping to get the time to factor this into the solver but couldn't... I think what we can do short-term is to run the pieces of `typeCheckExpression` here individually instead and disable diagnostics while running the solver for the first time - if PreCheckExpression fails - the whole method fails because there is something structural going on, if first solve fails - fallback to opening just like today.